File: src/com/timepath/curses/Terminal.java
package com.timepath.curses;

import static com.timepath.curses.Terminal.FONT_SIZE;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.FontMetrics;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.RenderingHints;
import java.awt.Toolkit;
import java.awt.geom.AffineTransform;
import java.util.Arrays;
import java.util.logging.Logger;
import javax.swing.JComponent;

@SuppressWarnings("serial")
public class Terminal extends JComponent {

    private static final Logger LOG = Logger.getLogger(Terminal.class.getName());

    public int xPos, yPos;
    public int w, h;
    public Color[] bgBuf, fgBuf;
    public char[] charBuf;
    public Dimension m;

    public Point caret = new Point(0, 0);

    public static final int FONT_SIZE = 12;

    /**
     * Java2D assumes 72 DPI.
     */
    public Font f = new Font(Font.MONOSPACED, Font.PLAIN, (int) Math.round(FONT_SIZE * Toolkit.getDefaultToolkit().getScreenResolution() / 72.0));
    public FontMetrics fm = this.getFontMetrics(f);

    protected Terminal() {
        super();
        m = new Dimension(fm.stringWidth(" "), fm.getHeight() - fm.getLeading());
    }

    public Terminal(int w, int h) {
        this();
        this.w = w;
        this.h = h;
        charBuf = new char[w * h];
        bgBuf = new Color[w * h];
        fgBuf = new Color[w * h];
        clear();
    }

    public void clear() {
        Arrays.fill(charBuf, (char) 0);
        Arrays.fill(bgBuf, Color.BLACK);
        Arrays.fill(fgBuf, Color.WHITE);
    }

    @Override
    public Dimension getPreferredSize() {
        if (m == null) {
            return super.getPreferredSize();
        }
        return new Dimension(w * m.width, h * m.height);
    }

    @Override
    public Dimension getMinimumSize() {
        return getPreferredSize();
    }

    @Override
    public void paint(Graphics graphics) {
        Graphics2D g = (Graphics2D) graphics;

        Color oldColor = g.getColor();
        AffineTransform oldAt = g.getTransform();
        AffineTransform newAt = new AffineTransform();
        newAt.translate(xPos * m.width, yPos * m.height);
        g.transform(newAt);

        g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_GASP);
        g.setFont(f);

        char character;
        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                Rectangle r = new Rectangle(x * m.width, y * m.height, m.width, m.height);
                g.setColor(bgBuf[x + y * w]);
                g.fillRect(r.x, r.y, r.width, r.height);
                g.setColor(fgBuf[x + y * w]);
                if ((character = charBuf[x + y * w]) == 0) {
                    continue;
                }
                g.drawString(String.valueOf(character), r.x, r.y + fm.getAscent());
            }
        }

        g.setTransform(oldAt);
        g.setColor(oldColor);
    }

    public void position(int x, int y) {
        caret.setLocation(x, y);
    }
    
    public void write(Object o) {
        write(String.valueOf(o));
    }

    public void write(String text) {
        char[] chars = text.toCharArray();
        int idx;
        for (int i = 0; i < chars.length; i++) {
            idx = (caret.x + i) + (caret.y * w);
            if (idx >= 0 && idx < charBuf.length) {
                charBuf[idx] = chars[i];
            }
        }
    }

    public Point cellToView(long ptr) {
        long x = ptr % w;
        long y = ptr / w;
        Point p = new Point((int) (x * m.width), (int) (y * m.height));
        p.translate(xPos * m.width, yPos * m.height);
        return p;
    }

    public int viewToCell(Point p) {
        p.translate(-xPos * m.width, -yPos * m.height);
        if (p.x < 0 || p.x >= w * m.width) {
            return -1;
        }
        if (p.y < 0 || p.y >= h * m.height) {
            return -1;
        }
        int x = p.x / m.width;
        int y = p.y / m.height;
        return (w * y) + x;
    }
}

File: src/main/kotlin/com/timepath/curses/Terminal.kt
package com.timepath.curses

import java.awt.*
import java.awt.geom.AffineTransform
import javax.swing.JComponent

public open class Terminal(w: Int = 0, h: Int = 0) : JComponent() {

    private val FONT_SIZE = 12
    /** Java2D assumes 72 DPI. */
    private val termFont = Font(Font.MONOSPACED, Font.PLAIN,
            Math.round((FONT_SIZE * Toolkit.getDefaultToolkit().getScreenResolution()).toDouble() / 72.0).toInt())
    private val fontMetrics = getFontMetrics(termFont)
    public var xPos: Int = 0
    public var yPos: Int = 0
    public var bgBuf: Array<Color> = Array(w * h) { Color.BLACK }
    public var fgBuf: Array<Color> = Array(w * h) { Color.WHITE }
    protected var metrics: Dimension = Dimension(fontMetrics.stringWidth(" "),
            fontMetrics.getHeight() - fontMetrics.getLeading())
    var termWidth: Int = w
    var termHeight: Int = h
    var charBuf: CharArray = CharArray(w * h)
    private val caret = Point(0, 0)

    public fun clear() {
        charBuf.fill(0.toChar())
        bgBuf.fill(Color.BLACK)
        fgBuf.fill(Color.WHITE)
    }

    override fun paint(g: Graphics) = (g as Graphics2D).let { g ->
        val oldColor = g.getColor()
        val oldAt = g.getTransform()
        val newAt = AffineTransform()
        newAt.translate((xPos * metrics.width).toDouble(), (yPos * metrics.height).toDouble())
        g.transform(newAt)
        g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_GASP)
        g.setFont(termFont)
        for (y in termHeight.indices) {
            for (x in termWidth.indices) {
                val px = x * metrics.width
                val py = y * metrics.height
                g.setColor(bgBuf[x + (y * termWidth)])
                g.fillRect(px, py, metrics.width, metrics.height)
                g.setColor(fgBuf[x + (y * termWidth)])
                val character = charBuf[x + (y * termWidth)]
                if (character.toInt() == 0) {
                    continue
                }
                g.drawString(character.toString(), px, py + fontMetrics.getAscent())
            }
        }
        g.setTransform(oldAt)
        g.setColor(oldColor)
    }

    override fun getPreferredSize() = Dimension(termWidth * metrics.width, termHeight * metrics.height)

    override fun getMinimumSize() = getPreferredSize()

    public fun position(x: Int, y: Int): Unit = caret.setLocation(x, y)

    public fun write(o: Any): Unit = o.toString().forEachIndexed { i, c ->
        val idx = caret.x + i + (caret.y * termWidth)
        if ((idx >= 0) && (idx < charBuf.size())) {
            charBuf[idx] = c
        }
    }

    public fun cellToView(ptr: Long): Point {
        val x = ptr % termWidth
        val y = ptr / termWidth
        return Point(
                ((x * metrics.width) + (xPos * metrics.width)).toInt(),
                ((y * metrics.height) + (yPos * metrics.height)).toInt()
        )
    }

    public fun viewToCell(p: Point): Int {
        p.translate(-xPos.times(metrics.width), -yPos * metrics.height)
        if ((p.x < 0) || (p.x >= (termWidth * metrics.width))) {
            return -1
        }
        if ((p.y < 0) || (p.y >= (termHeight * metrics.height))) {
            return -1
        }
        val x = p.x / metrics.width
        val y = p.y / metrics.height
        return (termWidth * y) + x
    }

}
